### PR TITLE
[perf] Cause RollupService to verify DML needed

### LINKED ISSUE #1269: Rollup should only execute DML if it would change the parent's value
**Is your feature request related to a problem? Please describe.**
If a child record is modified in a way that the rollup config determines it _might_ change the parent value, then the rollup is executed and the parent is updated. This doesn't verify that the parent value needs to be changed.

**Describe the solution you'd like**
When a child record is modified and that modification is a field that triggers a rollup calculate, DLRS should execute the aggregate query then query the parent to ensure the DML is necessary. Any parent's already holding the correct values should be excluded from the DML.

**Describe alternatives you've considered**
We could make DLRS understand how to parse SOQL WHERE clauses to better understand record changes so as to better determine when a rollup should and should not execute. I really don't want to do this and it has nearly infinite complexity.

**Additional context**
In my opinion this is one of the key causes for a belief that DLRS is inefficient. Causing DML on records that might not need it creates cascading sub-transactions and opens us the risk of triggering ever more automation and updating more records. I believe that if we can fix this it will make a significant improvement on reports of "record lock" "CPU Timeout" and other performance-based problems, including things like Scheduled actions consuming the entire 10K DML writes during large loads on heavily rollup-up objects.

### PR DESCRIPTION
**Pull request by aheber**
Added specific tests to validate these features on both realtime and scheduled mode jobs.
Other tests updates reflect behavior differences in limit counts due to this feature.

# Critical Changes

# Changes

# Issues Closed
#1269

### CODE REVIEW

**Comment by aheber** [comment]
Wait on this one, I still need to add a setting to disable the "should I DML" verification.

**Comment by aheber** [comment]
This is ready for review!
Haven't looked at why the CI is failing.

File: dlrs/main/classes/RollupServiceTest.cls
Comment (Szandor72) [inline dlrs/main/classes/RollupServiceTest.cls:1046]: This adds up. One + One + Two + Two + One + One = 8

File: dlrs/main/classes/RollupServiceTest.cls
Comment (Szandor72) [inline dlrs/main/classes/RollupServiceTest.cls:1068]: This one doesn't add up, so I am confused, why we need 2 more queries than before.

File: dlrs/main/classes/RollupServiceTest.cls
Comment (Szandor72) [inline dlrs/main/classes/RollupServiceTest.cls:1074]: Same, 2 more than before

**Review by Szandor72 — COMMENTED** [review]
Were making long methods a tiny bit longer and the logic is still tight. It would have helped me (and will maybe help future maintainers) to have one or two actual unit tests for the new helper methods. 

I think if we indicate how many query or rows are used in comments, the comments should match the assert values. Would it be possible to achieve this without headaches (i.e. investigation/ debug log reading)?

Both are just quality of life things, no blockers. Would you like to address any of them @aheber  ? 

Otherwise we can move forward after you had a look on the [deployment errors](https://github.com/SFDO-Community/declarative-lookup-rollup-summaries/actions/runs/3953718165/jobs/6770297857).

File: dlrs/main/classes/RollupServiceTest.cls
Comment (aheber) [reply dlrs/main/classes/RollupServiceTest.cls:1068]: 7 increased to 8 on line 1046 which means line 1068 now has a base +1 on it's math, then we added another query which is another +1, totaling in +2 being added though only one was caused by line 1062.

File: dlrs/main/classes/RollupServiceTest.cls
Comment (aheber) [reply dlrs/main/classes/RollupServiceTest.cls:1074]: Same comment as above, inherited additional base count.

**Review by Szandor72 — APPROVED** [review]
Thanks for extracting into subclass and adding dedicated tests. Moving forward, I want to campaign for assert messages :)

**Comment by Szandor72** [comment]
@Riley-Schill Would you like to do some QE before merging?

**Pull request merged**
